src/hyperion/external_interaction/callbacks/plan_reactive_callback.py

```diff
@@ -26,29 +29,52 @@ class will be activated, and on recieving the stop document for the
         override 'start' etc. should include a call to super().start(...) etc.
         The logic of how activation is triggered will change to a more readable, version
         in the future (https://github.com/DiamondLightSource/hyperion/issues/964)."""
+
         super().__init__(emit=emit)
         self.active = False
         self.activity_uid = 0
+        self.log = log  # type: ignore # this is initialised to None and not annotated in the superclass
+
+    def _run_activity_gated(self, func, doc, override=False):
+        running_gated_function = override or self.active
+        if not running_gated_function:
+            return doc
+        if self.log:
+            try:
+                return func(doc)
+            except Exception as e:
+                self.log.exception(e)
+                raise
+        else:
+            return func(doc)
 
     def start(self, doc: RunStart) -> RunStart | None:
         callbacks_to_activate = doc.get("activate_callbacks")
         if callbacks_to_activate:
-            self.active = type(self).__name__ in callbacks_to_activate
+            activate = type(self).__name__ in callbacks_to_activate
+            self.active = activate
+            self.optional_info_log(
+                f"{'' if activate else 'not'} activating {type(self).__name__}"
+            )
             self.activity_uid = doc.get("uid")
-        return self.activity_gated_start(doc) if self.active else doc
+        return self._run_activity_gated(self.activity_gated_start, doc)
 
     def descriptor(self, doc: EventDescriptor) -> EventDescriptor | None:
-        return self.activity_gated_descriptor(doc) if self.active else doc
+        return self._run_activity_gated(self.activity_gated_descriptor, doc)
 
     def event(self, doc: Event) -> Event | None:
-        return self.activity_gated_event(doc) if self.active else doc
+        return self._run_activity_gated(self.activity_gated_event, doc)
 
     def stop(self, doc: RunStop) -> RunStop | None:
         do_stop = self.active
         if doc.get("run_start") == self.activity_uid:
             self.active = False
             self.activity_uid = 0
-        return self.activity_gated_stop(doc) if do_stop else doc
+        return (
+            self._run_activity_gated(self.activity_gated_stop, doc, override=True)
```

Comment:
The point of override is to be able to `return` the activity gated function (to pass on the document to any potential following consumers) - at the end of the run it gets called with `override=True` and `self.active=False` because `self.active` is set to `False` immediately after setting `do_stop`

Reply:
Oh yh, my bad. Was tired when I wrote this